src/main/java/org/gradle/profiler/asyncprofiler/AsyncProfilerDistribution.java

```diff
@@ -0,0 +1,156 @@
+package org.gradle.profiler.asyncprofiler;
+
+import org.gradle.profiler.CommandExec;
+import org.gradle.profiler.OperatingSystem;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Model an async-profiler distribution, from 3+.
+ * <p>
+ * Note, that Async-Profiler executable was as shell script up to version 2.x, also
+ * the async profile library had the `.so` extension even on macOS, (a .dylib symlink
+ * was introduced in 2.8). Starting from async-profiler 3.0, the executable is a binary.
+ * </p>
+ *
+ * <p>
+ * <pre><code>
+ * async-profiler-3.0-macos
+ * ├── bin
+ * │   └── asprof
+ * ├── CHANGELOG.md
+ * ├── lib
+ * │   ├── async-profiler.jar
+ * │   ├── converter.jar
+ * │   └── libasyncProfiler.dylib
+ * ├── LICENSE
+ * └── README.md
+ * </code></pre>
+ *
+ * <pre><code>
+ * async-profiler-4.2-macos
+ * ├── bin
+ * │   ├── asprof
+ * │   └── jfrconv
+ * ├── include
+ * │   └── asprof.h
+ * ├── lib
+ * │   └── libasyncProfiler.dylib
+ * ├── LICENSE
+ * └── README.md
+ * </code></pre>
+ * </p>
+ */
+public class AsyncProfilerDistribution {
+    private final String source;
+    private final File executable;
+    private final File library;
+    private final Version version;
+
+    AsyncProfilerDistribution(File home, String source) {
+        this.source = source;
+
+        // Handles async-profiler 3.0+
+        File asprofBinary = new File(home, "bin/asprof");
+        if (asprofBinary.isFile()) {
+            executable = asprofBinary;
+            String libExtension;
+            // if async profiler 3+
+            if (OperatingSystem.isLinuxX86()) {
+                libExtension = ".so";
+            } else if (OperatingSystem.isMacOS()) {
+                libExtension = ".dylib";
+            } else {
+                throw new IllegalStateException("Unsupported operating system: " + OperatingSystem.getId());
+            }
+
+            library = new File(home, "lib/libasyncProfiler" + libExtension);
+        } else {
+            // Understand async-profiler 2.x, so we can at least report the version used
+            executable = new File(home, "profiler.sh");
+            library = new File(home, "build/libasyncProfiler.so");
+        }
+        System.out.println(asprofBinary.getAbsolutePath()); // TODO remove
+        if (!executable.isFile() // || !Files.isExecutable(executable.toPath())
+            || !library.isFile() // || !Files.isExecutable(library.toPath())
+        ) {
+            throw new IllegalStateException("Invalid async-profiler distribution at: " + home);
+        }
+
+        try {
+            executable.setExecutable(true);
```

Comment:
💅 I would personally not set permissions on behalf of users, but rather check them. 
I would only set them if distribution that we unzip doesn't have permissions.
Do we need to do this?

Reply:
Unfortunately this was needed, because the java extractor do not set the executable bit (while command line extractor do it). However, your comment made me realize that I overlooked the code in `AsyncProfilerDownload` that sets permissions.